### PR TITLE
Adding terminal  user interface

### PR DESCRIPTION
**Pull request by Geo-7**
Fixes # Nothing
Terminal UI is a new feature.

- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)
Now it just can only login to server and shows active logins
and it is cleaning up expired sessions.
cargo clippy shows some problem in kanidm_proto/src/v1.rs, but I hadn't changed that file.

### CODE REVIEW

**Comment by Firstyear** [comment]
I think this should be behind a feature gate, since it's still in development, and there is a lot of other changes going on. Would you mind doing that? 

Another option could be that with the upcoming talk about reworks of the cli internals, we might end up breaking this, so there is a solid argument to make this a seperate repo/project that consumes the client libraries and you target stable versions instead of the dev versions? 

What do you think?

**Comment by Geo-7** [comment]
feature gate or being a separate project both of them are ok .
I've not used the kanidm client library to reduce the code size and duplication but maybe using libs would be better in a long run.
yes there would be lot of other changes, and TUI can only login now, I made this early pull request to choose a correct path at the beginning.

**Comment by Firstyear** [comment]
I think feature gate might be the way to go, but we just accept that it may not be updated straight away? How's that? 

I'll do the code review shortly though :)

File: kanidm_tools/src/cli/session.rs
Comment (Firstyear) [inline kanidm_tools/src/cli/session.rs:150]: I feel like the `pass: Option<>` is a bit weird, it might be better to have a way to have a trait or a call ack that defines a way to get the password, which can then hook back to the relevant class of ui.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: kanidm_tools/Cargo.toml
Comment (Geo-7) [reply kanidm_tools/Cargo.toml:46]: Ok,

File: kanidm_tools/src/cli/session.rs
Comment (Geo-7) [reply kanidm_tools/src/cli/session.rs:150]: Ok, I'll revert backs all change to cli libs and start using kanidm_client ...is it ok?

File: kanidm_tools/src/cli/session.rs
Comment (Geo-7) [reply kanidm_tools/src/cli/session.rs:463]: same as above, I could remove this and use  kanidm_client libs

Comment (Firstyear) [reply kanidm_tools/src/cli/session.rs:463]: Yeah, I think this is starting to look in favour of seperate tool then since the client libs are the contact point, and that avoids the need to feature gate.

Comment (Firstyear) [reply kanidm_tools/src/cli/session.rs:463]: IE perhaps start kanidm_tui in parallel to kanidm_tools?

Comment (Geo-7) [reply kanidm_tools/src/cli/session.rs:463]: sounds good to me, I would create kanidm_tui folder and remove all dependency to cli libs,

**Comment by yaleman** [comment]
I'd be wary of adding another UI to the **core** project while it's at such an early alpha stage - it's another thing to provide maintenance and support for, and builds user expectations of ongoing feature support/development.

While I appreciate it's cool to support every possible method of interaction,  every line of code is a magnet for technical debt and the CLI/WebUI  are already fairly complex and have a long way to go.

Perhaps this should be a standalone project leveraging the `kanidm_client` module as the method for interaction, providing an example of how to do such a thing?

**Comment by Geo-7** [comment]
> I'd be wary of adding another UI to the **core** project while it's at such an early alpha stage - it's another thing to provide maintenance and support for, and builds user expectations of ongoing feature support/development.
> 
> While I appreciate it's cool to support every possible method of interaction, every line of code is a magnet for technical debt and the CLI/WebUI are already fairly complex and have a long way to go.
> 
> Perhaps this should be a standalone project leveraging the `kanidm_client` module as the method for interaction, providing an example of how to do such a thing?
what if we make it as a lib for kanidm_tools and make it **optional**. do you think on that case still builds user expectations?

**Comment by yaleman** [comment]
Ultimately, the existence of a tool implies it'll work, so being optional's not exactly a salve - but yes, at this stage, it should be under a feature flag as requested earlier. 😄

**Comment by Firstyear** [comment]
> Ultimately, the existence of a tool implies it'll work, so being optional's not exactly a salve - but yes, at this stage, it should be under a feature flag as requested earlier. 😄

Well, if we make it kanidm_tui in parallel, we don't need to include it in the workspace, so it won't build by default. But it can be built and used manually within that workspace still. That way we avoid the feature flag, but can have it here "monorepo" style.

File: kanidm_tools/Cargo.toml
Comment (Geo-7) [reply kanidm_tools/Cargo.toml]: It is done, now kanidm_tui is a optional lib for kanidm_tools

File: kanidm_tools/src/cli/session.rs
Comment (Geo-7) [reply kanidm_tools/src/cli/session.rs]: It is done now TUI does not use cli session.

**Comment by Geo-7** [comment]
> > Ultimately, the existence of a tool implies it'll work, so being optional's not exactly a salve - but yes, at this stage, it should be under a feature flag as requested earlier. smile
> 
> Well, if we make it kanidm_tui in parallel, we don't need to include it in the workspace, so it won't build by default. But it can be built and used manually within that workspace still. That way we avoid the feature flag, but can have it here "monorepo" style.
> 
> But yes, perhaps having this as a seperate repo and targeting stable versions is actually the best option since we seem to flux/change the api alot.

Nice idea, I removed it from workspace.
for checking tui now we should use `cargo run --bin kanidm --features tui tui`

**Comment by Geo-7** [comment]
I did not close this, After I run command ```git push -f origin HEAD~2:tui``` on my local computer it was closed, Sorry I'll open new one, I'll remove unnecessary commits it would be cleaner.

**Comment by Geo-7** [comment]
> I did not close this, After I run command `git push -f origin HEAD~2:tui` on my local computer it was closed, Sorry I'll open new one, I'll remove unnecessary commits it would be cleaner.

Reason :I was in a kanidm master baranch

File: kanidm_tools/src/cli/session.rs
Comment (Geo-7) [reply kanidm_tools/src/cli/session.rs]: Password is not option anymore.

**Review by yaleman — CHANGES_REQUESTED** [review]
The `kanidm_proto 1.1.0-alpha.7` change `/Cargo.lock` seems like a problematic change - the version shouldn't have to be specified since it's a relative dependency/import.

File: kanidm_tui/Cargo.toml
Comment (Firstyear) [inline kanidm_tui/Cargo.toml:15]: As yaleman mentioned, since it's a relative import you should use path = " " instead.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: kanidm_tui/Cargo.toml
Comment (Geo-7) [reply kanidm_tui/Cargo.toml]: It is done.

**Comment by Geo-7** [comment]
> The `kanidm_proto 1.1.0-alpha.7` change in `/Cargo.lock` seems like a problematic change - the version shouldn't have to be specified since it's a relative dependency/import.

Neat, It's done.

**Comment by Geo-7** [comment]
I have another commit which break kanidm_tui into
```mod ui
mod app
mod messages```
Should I commit it here or wait and make another pull request after this one merged?
@Firstyear 
@yaleman

**Comment by yaleman** [comment]
Might as well include it here.

**Comment by Geo-7** [comment]
> Might as well include it here.
Ok I did it, It is now cleaner than single file.

**Comment by Firstyear** [comment]
I think having reviewed this again, and how much seems to be improving and changing rapidly, maybe the best way to help you unblock so you can achieve more, is to make this a seperate repo, and later we'll bring it under the umbrella once it calms a bit. What do you think? That way you can also target stable instead of the moving master versions.

**Comment by Geo-7** [comment]
I agree, it is changing a lot, Now I'll close this pull request.

**Comment by Firstyear** [comment]
> I agree, it is changing a lot, Now I'll close this pull request.

It's really cool though! So please keep it up! :)

**Comment by Geo-7** [comment]
> > I agree, it is changing a lot, Now I'll close this pull request.
> 
> It's really cool though! So please keep it up! :)

I will, thank you.

**Pull request closed without merging**
